kernel/comps/keyboard/src/i8042_chip/mouse.rs

```diff
@@ -0,0 +1,422 @@
+    spin::Mutex::new(MousePacketState::new());
+
+pub(super) fn init(controller: &mut I8042Controller) -> Result<(), I8042ControllerError> {
+    const DEVICE_ID_STANDARD: u8 = 0x00;
+    const DEVICE_ID_INTELLIMOUSE: u8 = 0x03;
+    const DEVICE_ID_INTELLIMOUSE_EXPLORER: u8 = 0x04;
+    const MOUSE_CMD_ENABLE_DATA_REPORTING: u8 = 0xF4;
+
+    // Reset mouse device by sending `PS2_CMD_RESET` to the second PS/2 port.
+    send_mouse_command(controller, PS2_CMD_RESET)?;
+
+    // The response should be `PS2_ACK` and `PS2_BAT_OK`, followed by the device PS/2 ID.
+    if controller.wait_and_recv_data()? != PS2_ACK {
+        return Err(I8042ControllerError::DeviceResetFailed);
+    }
+    // The reset command may take some time to finish. Try again a few times.
+    if (0..5).find_map(|_| controller.wait_and_recv_data().ok()) != Some(PS2_BAT_OK) {
+        return Err(I8042ControllerError::DeviceResetFailed);
+    }
+
+    let mut device_id = controller
+        .wait_and_recv_data()
+        .unwrap_or(DEVICE_ID_STANDARD);
+    log::info!("PS/2 mouse device ID: 0x{:02X}", device_id);
+
+    // Try to enable IntelliMouse by setting sample rates: 200, 100, 80.
+    // Reference: https://wiki.osdev.org/Mouse_Input
+    if enable_intellimouse(controller).is_ok() {
+        // Query device ID again.
+        if let Ok(new_id) = get_mouse_device_id(controller) {
+            device_id = new_id;
+            log::info!("PS/2 mouse upgraded device ID: 0x{:02X}", device_id);
+        }
+    }
+
+    // Enable data reporting.
+    send_mouse_command(controller, MOUSE_CMD_ENABLE_DATA_REPORTING)?;
+    if controller.wait_and_recv_data()? != PS2_ACK {
+        return Err(I8042ControllerError::DeviceResetFailed);
+    }
+
+    let mut irq_line = IrqLine::alloc()
+        .and_then(|irq_line| {
+            IRQ_CHIP
+                .get()
+                .unwrap()
+                .map_isa_pin_to(irq_line, ISA_INTR_NUM)
+        })
+        .map_err(|_| I8042ControllerError::DeviceAllocIrqFailed)?;
+    irq_line.on_active(handle_mouse_input);
+    IRQ_LINE.call_once(|| irq_line);
+
+    // Create and register the i8042 mouse device.
+    let mouse_device = Arc::new(I8042Mouse::new());
+    let registered_device = aster_input::register_device(mouse_device);
+    REGISTERED_DEVICE.call_once(|| registered_device);
+
+    // Configure packet length based on device ID.
+    if device_id == DEVICE_ID_INTELLIMOUSE || device_id == DEVICE_ID_INTELLIMOUSE_EXPLORER {
+        let mut state = MOUSE_PACKET_STATE.lock();
+        state.set_packet_len(PACKET_LEN_INTELLIMOUSE);
+    }
+
+    Ok(())
+}
+
+/// Sends a command to the mouse.
+fn send_mouse_command(
+    controller: &mut I8042Controller,
+    command: u8,
+) -> Result<(), I8042ControllerError> {
+    controller.write_to_second_port(command)?;
+    Ok(())
+}
+
+/// Sends a command with one argument to the mouse.
+fn send_mouse_command_with_arg(
+    controller: &mut I8042Controller,
```

Comment:
Also passing the controller in this way is not very elegant. How about defining a `MouseInitCtx` type:
```rust
struct MouseInitCtx(&mut I8042Controller);

impl MouseInitCtx {
   // ...
}
```
So we can write:
```rust
    let mut device_id = controller
        .wait_and_recv_data()?;
    log::info!("PS/2 mouse device ID: 0x{:02X}", device_id);

    let init_ctx = MouseInitCtx(controller);

    // Try to enable IntelliMouse.
    if device_id == DEVICE_ID_STANDARD && init_ctx.enable_intellimouse().is_ok() {
        device_id = init_ctx.get_device_id()?;
        log::info!("PS/2 mouse upgraded device ID: 0x{:02X}", device_id);
    }

    // Enable data reporting.
    init_ctx.enable_data_reporting();
```